Update CarSpawnSettings.cs

catolm opened the pull request:
Random amount of fuel between two integers.

Random engine parts, configurable chance for each part.
Random engine parts tiers, configurable individual chanc for high, medium, and low engine parts.

Engine parts got random condition.

In total, the change makes cars spawn with random fuel, random set of engine parts in where each individual part is of random tier/quality and condition.

catolm committed:
Update CarSpawnSettings.cs

Random amount of fuel between two integers.

Random engine parts. Configurable max tier, individual chance percentage for high, medium and low engine parts. Also a master chance percentage that an engine part is being spawned into the engine part slot. Rerolled for every slot, meaning you can get all combinations between 0 and All parts, as well as all combinations of all qualities.

Engine parts got random condition.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -77,16 +78,21 @@ private void MaybeRepairModules(ModularCar car)
-            var fuelAmount = PluginConfig.FuelAmount;
-            if (fuelAmount == 0) return;
+            var maxFuelAmount = PluginConfig.MaxFuelAmount;
+			var minFuelAmount = PluginConfig.MinFuelAmount;
```
Use spaces instead of tabs to be consistent with the rest of the file.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -113,17 +119,31 @@ private void AddPartsToEngineStorage(EngineStorage engineStorage, int desiredTie
             if (engineStorage.inventory == null) return;
 
+			var partchance = PluginConfig.EnginePartChance;
+			var t3_chance = PluginConfig.EnginePartsTier3Chance;
```
Use camelCase for local variables. Example: `tier3Chance`.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -113,17 +119,31 @@ private void AddPartsToEngineStorage(EngineStorage engineStorage, int desiredTie
+			var t2_chance = PluginConfig.EnginePartsTier2Chance;
+			var t1_chance = PluginConfig.EnginePartsTier1Chance;
+			var t_roll = UnityEngine.Random.Range(0, 100);
+			var curtier = desiredTier;
```
Avoid abbreviating variables names unless it is a very standard abbreviation like `min` or `max`. Most abbreviations serve no purpose but to make the code harder to understand. In this case, `currentTier` is perfectly fine.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
+			[JsonProperty("MinFuelAmount")]
+            public int MinFuelAmount = 0;
 
-            [JsonProperty("FuelAmount")]
-            public int FuelAmount = 0;
```
Removing the `FuelAmount` option is a backwards-incompatible change, meaning I will need to increase the version number to 2.0 and ask people to update their configs if they don't want the new default behavior. I prefer making changes backwards compatible if possible to make updating easy.

My recommendation is to keep this option, use it instead of the min/max options if non-zero, and to hide it from the config if 0.

You can change this to `[JsonProperty("FuelAmount", DefaultValueHandling = DefaultValueHandling.Ignore)]` to cause the option to not be shown in the config when 0. I will make another change to the config loading on top of this before releasing which will automatically write to the config when there are changes or you won't necessarily notice the hiding behavior.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
         {
             [JsonProperty("EnginePartsTier")]
-            public int EnginePartsTier = 0;
+            public int EnginePartsTier = 3;
```
Change this default back to 0.

Similar to my comment about fuel, we should continue supporting this option and use it if non-zero, and hide it from the config if 0.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
 
-            [JsonProperty("FuelAmount")]
-            public int FuelAmount = 0;
+			[JsonProperty("EnginePartChance")]
+            public int EnginePartChance = 25;
```
The `EnginePartChance` option seems redundant with the `EnginePartsTier*Chance` options.

There are two ways to design this.
1. The tier of all parts for a given engine is randomly selected but consistent for the whole engine - this is the way you implemented it
2. The tier of each engine part is random, resulting in mixed tiers of parts within each engine

Do you have strong reasons for choosing the first option as opposed to the second? If you wanted to avoid mixed engine part quality, your implementation only avoids that per engine module, not per car since there can be multiple engine modules and each one can end up with a different quality parts.

IMO it's more realistic to randomize each part, so my suggestion would be to remove the `EnginePartChance` option and simply use the `EnginePartsTier*Chance` options. This will also be simpler for users to understand.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
-            public int EnginePartsTier = 0;
+            public int EnginePartsTier = 3;
+
+			[JsonProperty("EnginePartsTier3Chance")]
+            public int EnginePartsTier3Chance = 5;
+
+			[JsonProperty("EnginePartsTier2Chance")]
+            public int EnginePartsTier2Chance = 20;
+
+			[JsonProperty("EnginePartsTier1Chance")]
+            public int EnginePartsTier1Chance = 75;
```
Change all of these to 0 to match vanilla since this is not the only feature of this plugin. I understand it's probably one of the most used features at the moment, but that could certainly change when I release a new plugin that allows engines to work without parts, and when I update this plugin to allow choosing the modules that spawn randomly on cars. Both of those updates are pending an Oxide update because I needed to implement new hooks.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
+            public int EnginePartsTier1Chance = 75;
+
+            [JsonProperty("MaxFuelAmount")]
+            public int MaxFuelAmount = 11;
```
Change this default to 0 to match vanilla, and to be backwards compatible with users who currently have `FuelAmount` at 0. We don't want to force users to have fuel without opting in or being notified that the change is backwards incompatible (and a backwards incompatible change here is unnecessary).

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -113,17 +119,31 @@ private void AddPartsToEngineStorage(EngineStorage engineStorage, int desiredTie
+				t_roll = UnityEngine.Random.Range(0, 100);
                 if (item == null)
-                    TryAddEngineItem(engineStorage, i, desiredTier);
+					if(t_roll < t3_chance){
```
Use consistent spacing with the rest of the file.

```csharp
if (t_roll < t3_chance) {
```

WheteThunger reviewed (changes requested):
Thanks for your contribution! I have requested several changes. To summarize:

1. Follow existing conventions around spacing and naming.
2. Update config defaults to be non-intrusive and vanilla-friendly.
3. Make this change backwards compatible by supporting existing options if they are non-zero.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -132,8 +152,12 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
 
-            item.condition = component.condition.max;
-            item.MoveToContainer(engineStorage.inventory, slot, allowStack: false);
+            item.condition = UnityEngine.Random.Range(1.0f, 100.0f);
```
I missed this before. Engine part condition definitely needs to be configurable (min, max). Some people do not want engine parts to have condition so they will opt to use https://umod.org/plugins/engine-parts-durability to prevent them from taking damage, so they'll likely want to force engine part condition to 100 with this plugin.

catolm commented:
Thanks for the reply. Will update most of it.

A comment on EnginePartChance however.
As it is currently implemented together with the Tier chances and if you 
take a look at the call stack you will see that I did in fact make it so 
all tier parts are of a random mixture and that the EnginePartChance 
does not affect TierChance. It however affects the number of parts 
inside each engine, resulting in all possible combinations between a 
fully stacked engine and a totally empty engine. Whenever this integer 
is set at 25, it will on average only populate 25% of all engine parts 
slots in an engine. If at 100, it will ensure that all engines will 
spawn with parts in all of its slots and not just some. Likewise with 0, 
causing it to never spawn with parts at all. Currently running it on my 
server and it works like a charm and one have to scavenge multiple cars 
or craft parts to get a fully stocked engine. :=)

Summarize:
EnginePartTierChance only affects the chances a part will become a 
spesific tier.
EnginePartChance only affects how many of the engine slots that will be 
filled with parts.

Den 2020-10-22 22:53, skrev WhiteThunder:
> @WheteThunger requested changes on this pull request.
> 
> Thanks for your contribution! I have requested several changes. To
> summarize:
> 
>  	* Follow existing conventions around spacing and naming.
>  	* Update config defaults to be non-intrusive and vanilla-friendly.
>  	* Make this change backwards compatible by supporting existing
> options if they are non-zero.
> 
> -------------------------
> 
> In CarSpawnSettings.cs [1]:
> 
>> @@ -77,16 +78,21 @@ private void MaybeRepairModules(ModularCar car)
> 
>          private void MaybeAddFuel(ModularCar car)
>          {
> -            var fuelAmount = PluginConfig.FuelAmount;
> -            if (fuelAmount == 0) return;
> +            var maxFuelAmount = PluginConfig.MaxFuelAmount;
> +            var minFuelAmount = PluginConfig.MinFuelAmount;
> 
> Use spaces instead of tabs to be consistent with the rest of the file.
> 
> -------------------------
> 
> In CarSpawnSettings.cs [2]:
> 
>> @@ -113,17 +119,31 @@ private void
> AddPartsToEngineStorage(EngineStorage engineStorage, int desiredTie
>          {
>              if (engineStorage.inventory == null) return;
> 
> +            var partchance = PluginConfig.EnginePartChance;
> +            var t3_chance = PluginConfig.EnginePartsTier3Chance;
> 
> Use camelCase for local variables. Example: tier3Chance.
> -------------------------
> 
> In CarSpawnSettings.cs [3]:
> 
>> @@ -113,17 +119,31 @@ private void
> AddPartsToEngineStorage(EngineStorage engineStorage, int desiredTie
>          {
>              if (engineStorage.inventory == null) return;
> 
> +            var partchance = PluginConfig.EnginePartChance;
> +            var t3_chance = PluginConfig.EnginePartsTier3Chance;
> +            var t2_chance = PluginConfig.EnginePartsTier2Chance;
> +            var t1_chance = PluginConfig.EnginePartsTier1Chance;
> +            var t_roll = UnityEngine.Random.Range(0, 100);
> +            var curtier = desiredTier;
> 
> Avoid abbreviating variables names unless it is a very standard
> abbreviation like min or max. Most abbreviations serve no purpose but
> to make the code harder to understand. In this case, currentTier is
> perfectly fine.
> -------------------------
> 
> In CarSpawnSettings.cs [4]:
> 
>> -            [JsonProperty("FuelAmount")]
> -            public int FuelAmount = 0;
> 
> Removing the FuelAmount option is a backwards-incompatible change,
> meaning I will need to increase the version number to 2.0 and ask
> people to update their configs if they don't want the new default
> behavior. I prefer making changes backwards compatible if possible to
> make updating easy.
> 
> My recommendation is to keep this option, use it instead of the
> min/max options if non-zero, and to hide it from the config if 0.
> 
> You can change this to [JsonProperty("FuelAmount",
> DefaultValueHandling = DefaultValueHandling.Ignore)] to cause the
> option to not be shown in the config when 0. I will make another
> change to the config loading on top of this before releasing which
> will automatically write to the config when there are changes or you
> won't necessarily notice the hiding behavior.
> -------------------------
> 
> In CarSpawnSettings.cs [5]:
> 
>> @@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage
> engineStorage, int slot, int tier)
>          internal class CarSpawnSettingsConfig
>          {
>              [JsonProperty("EnginePartsTier")]
> -            public int EnginePartsTier = 0;
> +            public int EnginePartsTier = 3;
> 
> Change this default back to 0.
> 
> Similar to my comment about fuel, we should continue supporting this
> option and use it if non-zero, and hide it from the config if 0.
> -------------------------
> 
> In CarSpawnSettings.cs [6]:
> 
>> +            [JsonProperty("EnginePartChance")]
> +            public int EnginePartChance = 25;
> 
> The EnginePartChance option seems redundant with the
> EnginePartsTier*Chance options.
> 
> There are two ways to design this.
> 
>  	* The tier of all parts for a given engine is randomly selected but
> consistent for the whole engine - this is the way you implemented it
>  	* The tier of each engine part is random, resulting in mixed tiers
> of parts within each engine
> 
> Do you have strong reasons for choosing the first option as opposed to
> the second? If you wanted to avoid mixed engine part quality, your
> implementation only avoids that per engine module, not per car since
> there can be multiple engine modules and each one can end up with a
> different quality parts.
> 
> IMO it's more realistic to randomize each part, so my suggestion would
> be to remove the EnginePartChance option and simply use the
> EnginePartsTier*Chance options. This will also be simpler for users to
> understand.
> -------------------------
> 
> In CarSpawnSettings.cs [7]:
> 
>> +            [JsonProperty("EnginePartsTier3Chance")]
> +            public int EnginePartsTier3Chance = 5;
> +
> +            [JsonProperty("EnginePartsTier2Chance")]
> +            public int EnginePartsTier2Chance = 20;
> +
> +            [JsonProperty("EnginePartsTier1Chance")]
> +            public int EnginePartsTier1Chance = 75;
> 
> Change all of these to 0 to match vanilla since this is not the only
> feature of this plugin. I understand it's probably one of the most
> used features at the moment, but that could certainly change when I
> release a new plugin that allows engines to work without parts, and
> when I update this plugin to allow choosing the modules that spawn
> randomly on cars. Both of those updates are pending an Oxide update
> because I needed to implement new hooks.
> -------------------------
> 
> In CarSpawnSettings.cs [8]:
> 
>> @@ -147,10 +171,25 @@ private bool TryAddEngineItem(EngineStorage
> engineStorage, int slot, int tier)
>          internal class CarSpawnSettingsConfig
>          {
>              [JsonProperty("EnginePartsTier")]
> -            public int EnginePartsTier = 0;
> +            public int EnginePartsTier = 3;
> +
> +            [JsonProperty("EnginePartsTier3Chance")]
> +            public int EnginePartsTier3Chance = 5;
> +
> +            [JsonProperty("EnginePartsTier2Chance")]
> +            public int EnginePartsTier2Chance = 20;
> +
> +            [JsonProperty("EnginePartsTier1Chance")]
> +            public int EnginePartsTier1Chance = 75;
> +
> +            [JsonProperty("MaxFuelAmount")]
> +            public int MaxFuelAmount = 11;
> 
> Change this default to 0 to match vanilla, and to be backwards
> compatible with users who currently have FuelAmount at 0. We don't
> want to force users to have fuel without opting in or being notified
> that the change is backwards incompatible (and a backwards
> incompatible change here is unnecessary).
> -------------------------
> 
> In CarSpawnSettings.cs [9]:
> 
>>                  if (item == null)
> -                    TryAddEngineItem(engineStorage, i, desiredTier);
> +                    if(t_roll < t3_chance){
> 
> Use consistent spacing with the rest of the file.
> 
> if (t_roll < t3_chance) {
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [10], or unsubscribe
> [11].
> 
> Links:
> ------
> [1] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510420243
> [2] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510422643
> [3] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510423023
> [4] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510431135
> [5] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510435086
> [6] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510438290
> [7] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510438957
> [8] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510440491
> [9] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510441620
> [10] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#pullrequestreview-515071698
> [11] 
> https://github.com/notifications/unsubscribe-auth/ABICA3HCBQOT2UZ3T3FGG33SMCLS5ANCNFSM4S2DN3JA

catolm committed:
What you asked for, and more.

Changes you asked for, plus a explaination and renaming of something you misinterpeted.

It is infact random number of parts in each engine (except if configed to 100 or 0 for always all or always none) and each part has a random quality depending on the probability configured. ;)

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -77,16 +78,21 @@ private void MaybeRepairModules(ModularCar car)
+            if (maxFuelAmount < 0)
+                maxFuelAmount = fuelContainer.allowedItem.stackable;
+
+            if (minFuelAmount < 0)
+                minFuelAmount = 0;
```
I think we can safely remove this extra check since we already check `== 0` and `< 0` above. If the max stack size is 0 then low grade fuel isn't allowed which seems unlikely... unless the server is running some kind of alternative fuel mod? In which case they should probably not use this option in the first place.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -112,18 +118,43 @@ private void MaybeAddEngineParts(ModularCar car)
             if (engineStorage.inventory == null) return;
-
+			
+			/* */
```
Still seeing a number of formatting issues that I brought up in the previous revision. It makes it harder to review the code when the indentation is off. At a minimum, you can find/replace each tab with 4 spaces. Not sure what editor you are using, but usually IDEs nowadays will automatically conform to the conventions of the file being edited.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -112,18 +118,43 @@ private void MaybeAddEngineParts(ModularCar car)
+            var engineSlotFillChance = PluginConfig.EngineSlotFillChance; 
+			
+			// Tier chance is the chance for a an engine part to become the particular tier.
+			// Also rolled for every part individually, making it possible to get all possible combinations of qualities.
```
I don't think we need any comments describing how the chances work. Should be intuitive enough for experienced programmers. We can add plugin documentation to describe to users how these work.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -112,18 +118,43 @@ private void MaybeAddEngineParts(ModularCar car)
+			// - Together does these four variables make it so that engines can spawn with random number of parts with a random quality where you set the chances in config.
+			/* */
+			
+            var tierRoll = UnityEngine.Random.Range(0, 100);
```
Can this just be declared inside the for loop? Doesn't seem this initial value is used.

WheteThunger replied:
Still not seeing why we can't just do `tierRoll = UnityEngine.Random.Range(0, 100);` inside the for loop. The value is overriden each loop iteration, and the value is never used outside of the loop.

Similarly, `currentTier` can be declared in the loop as well.

The `desiredTier` parameter is still being ignored. We should be using it if non-zero for backwards compatibility. That logic can be in the for loop also.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -149,9 +189,30 @@ internal class CarSpawnSettingsConfig
+            public float EnginePartMaxCondition = 100.0f;
+			
+			[JsonProperty("EnginePartMinCondition")]
+            public float EnginePartMinCondition = 1.0f;
```
Change to 100 to be backwards compatible since currently all parts are spawned at 100% condition.

WheteThunger commented on CarSpawnSettings.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "EnginePartsTier": 3,
+  "EnginePartsTier3Chance": 5,
+  "EnginePartsTier2Chance": 20,
+  "EnginePartsTier1Chance": 75,
+  "MaxFuelAmount": 11,
+  "FuelAmount": 0,
+  "EngineSlotFillChance": 25,
+  "EnginePartMaxCondition": 100.0,
+  "EnginePartMinCondition": 1.0,
+  "HealthPercentage": -1.0,
+  "IncludeChassis": false,
+  "IncludeOwnedCars": false
+}
```
Not sure why this file was added since people won't see it.

Planning to add a "starter config" section in the documentation for people to adopt a randomization preset.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -132,8 +163,17 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
-            item.condition = component.condition.max;
-            item.MoveToContainer(engineStorage.inventory, slot, allowStack: false);
+			var minCondition = PluginConfig.EnginePartMinCondition; //(%)
+			var maxCondition = PluginConfig.EnginePartMaxCondition; //(%)
```
We can just name these with a `Percent` suffix to avoid needing to add comments.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -132,8 +163,17 @@ private bool TryAddEngineItem(EngineStorage engineStorage, int slot, int tier)
+				maxCondition = maxCondition;
+			
+            item.condition = UnityEngine.Random.Range(minCondition, maxCondition);
+            if(UnityEngine.Random.Range(0, 100) < chance || minCondition >= 100.0f){
```
It appears we can create the item above and then sometimes don't use it, depending on the outcome of this check. In cases like this, we should call `item.Remove()` to clean up when the item is not used.

However, in this case, we should simply move the check to before the item is created to avoid making the server do unnecessary work. Preferably outside of this method. It should also suffice to simply remove this chance option as I have suggested.

Also, why are we checking `minCondition` here?

WheteThunger commented:
I hadn't looked very closely, so I must have missed that the way you implemented it already provides a mix of different engine part qualities.

I think my point still stands that the purpose of the `EnginePartChance` option can already easily be achieved by using the other three options. For example, the following settings will have an overall 25% of spawning an engine part.

```
Tier1Chance = 25
Tier2Chance = 15
Tier3Chance = 5
```

Having an extra option requires the user to do math to understand the actual chances. The only benefit to the extra option is that it allows a user to change the overall chance while maintaining their existing ratio, but that's optimizing for what is probably an infrequent configuration change on their part, so I don't think it's worth the extra complexity.

catolm commented:
Agree, this would create a cleaner code and easier logic.

Den 2020-10-23 18:25, skrev WhiteThunder:
> I hadn't looked very closely, so I must have missed that the way you
> implemented it already provides a mix of different engine part
> qualities.
> 
> I think my point still stands that the purpose of the EnginePartChance
> option can already easily be achieved by using the other three
> options. For example, the following settings will have an overall 25%
> of spawning an engine part.
> 
> Tier1Chance = 25
> Tier2Chance = 15
> Tier3Chance = 5
> 
> Having an extra option requires the user to do math to understand the
> actual chances. The only benefit to the extra option is that it allows
> a user to change the overall chance while maintaining their existing
> ratio, but that's optimizing for what is probably an infrequent
> configuration change on their part, so I don't think it's worth the
> extra code.
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or unsubscribe
> [2].
> 
> Links:
> ------
> [1] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#issuecomment-715443898
> [2] 
> https://github.com/notifications/unsubscribe-auth/ABICA3BXGXHI7EGQICBDXCTSMGVBPANCNFSM4S2DN3JA

catolm commented:
Probably. I am natively suspicious to implicit information in 
programming, since loads of vulnerabilities comes from that. Always keep 
the buffers filled and checked for valid content, unless the risk is 
taken (which is probably ok to do in a game's plugin). Also not sure if 
umod/oxide is precompiling the code or not. If it is real time 
interpreting the code, any statements without brackets (for cleaner 
code) put extra load on the tokenizer and slow the execution time and 
increase processor use.

Will try to sort the formatting issues. They are coming from Notepad++. 
I'll pop it into Atom and reformat it there instead.

The code comments I mainly did for you to explain what I did :=) We can 
remove it now.

"var tierRoll = UnityEngine.Random.Range(0, 100);"

tierRoll must be initiated before the loop, we can just set it to 0.
Don't think C# allows for initiation of empty variables. (ie var 
tierRoll; isnt' allowed, but var tierRoll = 0; is)

Den 2020-10-23 18:20, skrev WhiteThunder:
> @WheteThunger requested changes on this pull request.
> -------------------------
> 
> In CarSpawnSettings.cs [1]:
> 
>> +            if (minFuelAmount < 0)
> +                minFuelAmount = 0;
> 
> I think we can safely remove this extra check since we already check
> == 0 and < 0 above. If the max stack size is 0 then low grade fuel
> isn't allowed which seems unlikely... unless the server is running
> some kind of alternative fuel mod? In which case they should probably
> not use this option in the first place.
> -------------------------
> 
> In CarSpawnSettings.cs [2]:
> 
>> @@ -112,18 +118,43 @@ private void MaybeAddEngineParts(ModularCar
> car)
>          private void AddPartsToEngineStorage(EngineStorage
> engineStorage, int desiredTier)
>          {
>              if (engineStorage.inventory == null) return;
> -
> +
> +            /* */
> 
> Still seeing a number of formatting issues that I brought up in the
> previous revision. It makes it harder to review the code when the
> indentation is off. At a minimum, you can find/replace each tab with 4
> spaces. Not sure what editor you are using, but usually IDEs nowadays
> will automatically conform to the conventions of the file being
> edited.
> -------------------------
> 
> In CarSpawnSettings.cs [3]:
> 
>> +            // Tier chance is the chance for a an engine part to
> become the particular tier.
> +            // Also rolled for every part individually, making it
> possible to get all possible combinations of qualities.
> 
> I don't think we need any comments describing how the chances work.
> Should be intuitive enough for experienced programmers. We can add
> plugin documentation to describe to users how these work.
> -------------------------
> 
> In CarSpawnSettings.cs [4]:
> 
>> +            /* */
> +            // It is rolled once for every slot it iterates through,
> increasing the likelyhood for a slot to be populated.
> +            // Setting this to 100 (%) will guarantee that engine
> will get a part in all slots, 0 (%) guarantees none.
> +            // Setting this to I.e 50 (%) will cause on average that
> vehicles are spawned with half of its needed engine parts.
> +            var engineSlotFillChance =
> PluginConfig.EngineSlotFillChance;
> +
> +            // Tier chance is the chance for a an engine part to
> become the particular tier.
> +            // Also rolled for every part individually, making it
> possible to get all possible combinations of qualities.
> +            var tier3Chance = PluginConfig.EnginePartsTier3Chance;
> +            var tier2Chance = PluginConfig.EnginePartsTier2Chance;
> +            var tier1Chance = PluginConfig.EnginePartsTier1Chance;
> +
> +            // - Together does these four variables make it so that
> engines can spawn with random number of parts with a random quality
> where you set the chances in config.
> +            /* */
> +
> +            var tierRoll = UnityEngine.Random.Range(0, 100);
> 
> Can this just be declared inside the for loop? Doesn't seem this
> initial value is used.
> -------------------------
> 
> In CarSpawnSettings.cs [5]:
> 
>>              [JsonProperty("FuelAmount")]
>              public int FuelAmount = 0;
> 
> +            [JsonProperty("EngineSlotFillChance")]
> +            public int EngineSlotFillChance = 25;
> +
> +            [JsonProperty("EnginePartMaxCondition")]
> +            public float EnginePartMaxCondition = 100.0f;
> +
> +            [JsonProperty("EnginePartMinCondition")]
> +            public float EnginePartMinCondition = 1.0f;
> 
> Change to 100 to be backwards compatible since currently all parts are
> spawned at 100% condition.
> -------------------------
> 
> In CarSpawnSettings.json [6]:
> 
>> +{
> +  "EnginePartsTier": 3,
> +  "EnginePartsTier3Chance": 5,
> +  "EnginePartsTier2Chance": 20,
> +  "EnginePartsTier1Chance": 75,
> +  "MaxFuelAmount": 11,
> +  "FuelAmount": 0,
> +  "EngineSlotFillChance": 25,
> +  "EnginePartMaxCondition": 100.0,
> +  "EnginePartMinCondition": 1.0,
> +  "HealthPercentage": -1.0,
> +  "IncludeChassis": false,
> +  "IncludeOwnedCars": false
> +}
> 
> Not sure why this file was added since people won't see it.
> 
> Planning to add a "starter config" section in the documentation for
> people to adopt a randomization preset.
> -------------------------
> 
> In CarSpawnSettings.cs [7]:
> 
>> +            var minCondition = PluginConfig.EnginePartMinCondition;
> //(%)
> +            var maxCondition = PluginConfig.EnginePartMaxCondition;
> //(%)
> 
> We can just name these with a Percent suffix to avoid needing to add
> comments.
> -------------------------
> 
> In CarSpawnSettings.cs [8]:
> 
>> @@ -132,8 +163,17 @@ private bool TryAddEngineItem(EngineStorage
> engineStorage, int slot, int tier)
>              var item = ItemManager.Create(component);
>              if (item == null) return false;
> 
> -            item.condition = component.condition.max;
> -            item.MoveToContainer(engineStorage.inventory, slot,
> allowStack: false);
> +            var minCondition = PluginConfig.EnginePartMinCondition;
> //(%)
> +            var maxCondition = PluginConfig.EnginePartMaxCondition;
> //(%)
> +            if(minCondition > maxCondition)
> +                maxCondition = maxCondition;
> +
> +            item.condition = UnityEngine.Random.Range(minCondition,
> maxCondition);
> +            if(UnityEngine.Random.Range(0, 100) < chance ||
> minCondition >= 100.0f){
> 
> It appears we can create the item above and then sometimes don't use
> it, depending on the outcome of this check. In cases like this, we
> should call item.Remove() to clean up when the item is not used.
> 
> However, in this case, we should simply move the check to before the
> item is created to avoid making the server do unnecessary work.
> Preferably outside of this method. It should also suffice to simply
> remove this chance option as I have suggested.
> 
> Also, why are we checking minCondition here?
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [9], or unsubscribe
> [10].
> 
> Links:
> ------
> [1] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510973574
> [2] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510974360
> [3] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510975765
> [4] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510976386
> [5] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510977239
> [6] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510978062
> [7] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510978938
> [8] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r510984314
> [9] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#pullrequestreview-515789380
> [10] 
> https://github.com/notifications/unsubscribe-auth/ABICA3HHZ2B34632SECNSYTSMGUNFANCNFSM4S2DN3JA

WheteThunger commented:
Plugins are compiled when loaded. They are also fairly sandboxes so many vulnerabilities are avoided.

Code contains logic, else it would serve no purpose. Not sure how to avoid that without adding higher levels of abstraction, which the config essentially achieves, as long as it is understood, which the documentation helps with.

WheteThunger commented on CarSpawnSettings.cs:
```diff
@@ -112,18 +118,43 @@ private void MaybeAddEngineParts(ModularCar car)
+                tierRoll = UnityEngine.Random.Range(0, 100);
                 if (item == null)
-                    TryAddEngineItem(engineStorage, i, desiredTier);
+                    if(tierRoll < tier3Chance){
+                        currentTier = 3;
+					} else if(tierRoll < tier2Chance){
+                        currentTier = 2;
+					} else{
+                        currentTier = 1;
+					}
+                    TryAddEngineItem(engineStorage, i, currentTier, engineSlotFillChance);
```
We need braces around this. Otherwise, `TryAddEngineItem` will always be called which is not supposed to happen when `item == null`.

catolm commented:
Kinda redundant tho? Since
"if (item == null) return false;"
is at near top of TryAddEngineItem.

It is just a common security practice to define variables outside of 
loops so the variable can reuse the same space within the heap? Just for 
efficiency purposes and overwriting old results in memory, so they 
cannot be harvested as easily. Also saves memory handling time within 
the kernel ye? But sure, just a rust plugin anyway.

Den 2020-10-23 19:59, skrev WhiteThunder:
> @WheteThunger requested changes on this pull request.
> -------------------------
> 
> In CarSpawnSettings.cs [1]:
> 
>> +                    if(tierRoll < tier3Chance){
> +                        currentTier = 3;
> +                    } else if(tierRoll < tier2Chance){
> +                        currentTier = 2;
> +                    } else{
> +                        currentTier = 1;
> +                    }
> +                    TryAddEngineItem(engineStorage, i, currentTier,
> engineSlotFillChance);
> 
> We need braces around this. Otherwise, TryAddEngineItem will always be
> called which is not supposed to happen when item == null.
> -------------------------
> 
> In CarSpawnSettings.cs [2]:
> 
>> +            /* */
> +            // It is rolled once for every slot it iterates through,
> increasing the likelyhood for a slot to be populated.
> +            // Setting this to 100 (%) will guarantee that engine
> will get a part in all slots, 0 (%) guarantees none.
> +            // Setting this to I.e 50 (%) will cause on average that
> vehicles are spawned with half of its needed engine parts.
> +            var engineSlotFillChance =
> PluginConfig.EngineSlotFillChance;
> +
> +            // Tier chance is the chance for a an engine part to
> become the particular tier.
> +            // Also rolled for every part individually, making it
> possible to get all possible combinations of qualities.
> +            var tier3Chance = PluginConfig.EnginePartsTier3Chance;
> +            var tier2Chance = PluginConfig.EnginePartsTier2Chance;
> +            var tier1Chance = PluginConfig.EnginePartsTier1Chance;
> +
> +            // - Together does these four variables make it so that
> engines can spawn with random number of parts with a random quality
> where you set the chances in config.
> +            /* */
> +
> +            var tierRoll = UnityEngine.Random.Range(0, 100);
> 
> Still not seeing why we can't just do tierRoll =
> UnityEngine.Random.Range(0, 100); inside the for loop. The value is
> overriden each loop iteration, and the value is never used outside of
> the loop.
> 
> Similarly, currentTier can be declared in the loop as well.
> 
> The desiredTier parameter is still being ignored. We should be using
> it if non-zero for backwards compatibility. That logic can be in the
> for loop also.
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [3], or unsubscribe
> [4].
> 
> Links:
> ------
> [1] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r511050079
> [2] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#discussion_r511049687
> [3] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#pullrequestreview-515888130
> [4] 
> https://github.com/notifications/unsubscribe-auth/ABICA3CBIA7QBXEGHGKWT33SMHABRANCNFSM4S2DN3JA

WheteThunger commented:
Agreed on just a Rust plugin, but I'm not totally following the rest. Those kinds of security concerns should ideally not be leaking into code at this level. We should also avoid prematurely optimizing, prioritizing maintainability. This is also just a number so there is no heap allocation.

catolm committed:
Probabilities

Tier chance now affects both spawn rate of parts as well as the tier. Making the previous EnginePartChance redundant, ergo removed. Prettyfied code and fixed default backwards compability.

WheteThunger commented:
Closing since I've manually merged this with changes. There were some bugs and some changes I requested that you had not yet made. I also moved things around for organization. You can check it out on the develop branch.

https://github.com/WheteThunger/CarSpawnSettings/tree/develop

I might make some additional changes before releasing, such as nesting the engine part options like this:

```json
{
  "EngineParts": {
    "Tier1Chance": 0,
    "Tier2Chance": 0,
    "Tier3Chance": 0,
    "MinCondition": 100.0,
    "MaxCondition": 100.0
  },
  "MinFuelAmount": 0,
  "MaxFuelAmount": 0,
  "HealthPercentage": -1.0,
  "IncludeChassis": false,
  "IncludeOwnedCars": false
}
```

catolm commented:
Nice :)

Den 2020-10-24 11:08, skrev WhiteThunder:
> Closing since I've manually merged this with changes. There were some
> bugs and some changes I requested that you had not yet made. I also
> moved things around for organization. You can check it out on the
> develop branch.
> 
> https://github.com/WheteThunger/CarSpawnSettings/tree/develop
> 
> I might make some additional changes before releasing, such as nesting
> the engine part options like this:
> 
> {
>   "EngineParts": {
>     "Tier1Chance": 0,
>     "Tier2Chance": 0,
>     "Tier3Chance": 0,
>     "MinCondition": 100.0,
>     "MaxCondition": 100.0
>   },
>   "MinFuelAmount": 0,
>   "MaxFuelAmount": 0,
>   "HealthPercentage": -1.0,
>   "IncludeChassis": false,
>   "IncludeOwnedCars": false
> }
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or unsubscribe
> [2].
> 
> Links:
> ------
> [1] 
> https://github.com/WheteThunger/CarSpawnSettings/pull/1#issuecomment-715885147
> [2] 
> https://github.com/notifications/unsubscribe-auth/ABICA3FN57BWLJUNSMWTGODSMKKP7ANCNFSM4S2DN3JA
